File: Linked_list/linked_list.cpp
#include <iostream>
using namespace std;

class Node{
    public:
        int data;
        Node* next;
        // Default Constructor
        Node(){
            data = 0;
            next = NULL;
        }
        // Constructor
        Node(int data){
            this->data = data;
            this->next = NULL;
        }
};

class LinkedList{
    Node* head;
    public:
        // Default constructor
        LinkedList(){head=NULL};
        // Class member functions
        void insertNode(int);
        void print();
        void deleteNode(int);
};

void LinkedList::insertNode(int data){
    Node* newNode = new Node(data);
    if(head==NULL){
        head = newNode;
        return;
    }
    Node* temp = head;
    while(temp->next){
        temp = temp->next;
    }
    temp->next = newNode;
    return;
}
void LinkedList::print(){
    Node* temp = head;
    if (head==NULL){
        cout << "" << endl;
    }
    while(temp!=NULL){
        cout << temp->data << "->";
        temp = temp->next;
    }
}
void LinkedList::deleteNode(int nodeOffset){
    Node*temp1 = head, *temp2=NULL;
    if(head==NULL){
        cout << "Empty List." << endl;
        return;
    }
    int len = 0;
    while(temp1 != NULL){
        temp1 = temp1->next;
        len++;
    }
    if(len<nodeOffset){
        cout << "index out of range." << endl;
    }
    if(nodeOffset==1){
        head = head->next;
        delete temp1;
        return;
    }
    while(nodeOffset-->1){
        temp2 = temp1;
        temp1 = temp1->next;
    }
    temp2->next = temp1->next;
    delete temp1;
}

int main(){
    LinkedList list;
    // insert nodes
    list.insertNode(1);
    list.insertNode(2);
    list.insertNode(3);

    list.print();
}

File: Linked_list/linked_list.py
class Node:
    data = None
    next = None

    def __init__(self, data):
        self.data = data
        #self.next = None

    def __repr__(self):
        return "<Node data: %s>" % self.data
    

class LinkedList:
    def __init__(self):
        self.head = None

    def isEmpty(self):
        return self.head == None
    
    def size(self):
        current = self.head
        count = 0

        while current:
            count += 1
            current = current.next_node

        return count
    
    def add_data(self,data):
        new_node = Node(data)
        new_node.next_node = self.head.next_node
        self.head = new_node


    def __repr__(self):
        nodes = []
        current = self.head

        while current:
            if current is self.head:
                nodes.append("[Head: %s]"%current.data)
            if current.next == None:
                nodes.append("[Tail: %s]"%current.data)
            else:
                nodes.append("[%s]"%current.data)
            
            current = current.next_node

        return '->'.join(nodes)
    
    def search(self, key):
        current = self.head
        while current:
            if current.data == key: return current
            else: current = current.next_node
        return None
    
    def insert(self, data, index):
        if index == 0: self.add(data)
        if index > 0:
            new = Node(data)
            position = index
            current = self.head

            while position > 1:
                current = current.next_node
                position-=1

            prev = current
            next = current.next_node

            prev.next_node = new
            new.next_node = next


    def remove(self, key):
        current = self.head
        previous = None
        found = False

        while current and not found:
            if current.data == key and current == self.head:
                self.head = current.next_node
                found = True
            if current.data == key:
                prev.next_node = current.next_node
                found = True
            else:
                prev = current
                current = current.next_node

        #return current


    def node_at_index(self, index):
        if index == 0: return self.head
        else:
            current = self.head
            position = 0
            while position < index:
                current = current.next_node
                position += 1
            
            return current

        


